Source/JavaScriptCore/runtime/ArrayBuffer.cpp

```diff
@@ -326,6 +341,142 @@ void ArrayBuffer::notifyDetaching(VM& vm)
+        return nullptr;
+
+    size_t initialBytes = roundUpToMultipleOf<PageCount::pageSize>(sizeInBytes.value());
+    if (!initialBytes)
+        initialBytes = PageCount::pageSize; // Make sure malloc actually allocates something, but not too much. We use null to mean that the buffer is detached.
+    size_t maximumBytes = roundUpToMultipleOf<PageCount::pageSize>(maxByteLength);
+    if (!maximumBytes)
+        maximumBytes = PageCount::pageSize; // Make sure malloc actually allocates something, but not too much. We use null to mean that the buffer is detached.
+
+    bool done = tryAllocate(vm,
+        [&] () -> BufferMemoryResult::Kind {
+            return BufferMemoryManager::singleton().tryAllocatePhysicalBytes(initialBytes);
+        });
+    if (!done)
+        return nullptr;
+
+    char* slowMemory = nullptr;
+    tryAllocate(vm,
+        [&] () -> BufferMemoryResult::Kind {
+            auto result = BufferMemoryManager::singleton().tryAllocateGrowableBoundsCheckingMemory(maximumBytes);
+            slowMemory = bitwise_cast<char*>(result.basePtr);
+            return result.kind;
+        });
+    if (!slowMemory) {
+        BufferMemoryManager::singleton().freePhysicalBytes(initialBytes);
+        return nullptr;
+    }
+
+    constexpr bool readable = false;
+    constexpr bool writable = false;
+    if (!OSAllocator::protect(slowMemory + initialBytes, maximumBytes - initialBytes, readable, writable)) {
+        dataLog("mprotect failed: ", safeStrerror(errno).data(), "\n");
+        RELEASE_ASSERT_NOT_REACHED();
+    }
+
+    auto handle = adoptRef(*new BufferMemoryHandle(slowMemory, initialBytes, maximumBytes, PageCount::fromBytes(initialBytes), PageCount::fromBytes(maximumBytes), MemorySharingMode::Shared, MemoryMode::BoundsChecking));
+    auto content = SharedArrayBufferContents::create(handle->memory(), sizeInBytes.value(), maxByteLength, WTFMove(handle), nullptr, SharedArrayBufferContents::Mode::Default);
+    return createShared(WTFMove(content));
+}
+
+Expected<void, GrowFailReason> SharedArrayBufferContents::grow(VM& vm, size_t newByteLength)
+{
+    if (!m_hasMaxByteLength)
+        return makeUnexpected(GrowFailReason::GrowSharedUnavailable);
+    ASSERT(m_memoryHandle);
+    return grow(Locker { m_memoryHandle->lock() }, vm, newByteLength);
+}
+
+Expected<void, GrowFailReason> SharedArrayBufferContents::grow(const AbstractLocker&, VM& vm, size_t newByteLength)
+{
+    // Keep in mind that newByteLength may not be page-size-aligned.
+    size_t sizeInBytes = m_sizeInBytes.load(std::memory_order_seq_cst);
+    if (sizeInBytes > newByteLength || m_maxByteLength < newByteLength)
+        return makeUnexpected(GrowFailReason::InvalidGrowSize);
+
+    if (sizeInBytes == newByteLength)
+        return { };
+
+    auto newPageCount = PageCount::fromBytesWithRoundUp(newByteLength);
+    auto oldPageCount = PageCount::fromBytes(m_memoryHandle->size()); // MemoryHandle's size is always page-size aligned.
```

Comment:
Do we have a TOCTOU issue with not pre-caching `m_memoryHandle->size()`?  We're using it at multiple places in this function.  Would it be better to pre-cache it in a local first?

Reply:
It does not happen since we already took a lock of this memory handle.

Reply:
Oh, the `const AbstractLocker&`.  Right!